### ISSUE TITLE
Add a Widget that gets its height from its children and allows its children to position themselves based on siblings.

### ISSUE THREAD

**Issue by Gilga89** [flutter/flutter#99538]
## Use case
I have two widgets, **A** and **B**. I want both to be siblings (same parent). Both widgets have an **unknown height** and by that I mean the height is dynamic, can be wrapped in an `InstrinsicHeight`, and the widget could be a `Text` for example. 

I need to position **B** center to the bottom of **A** like in the image below:
![](https://i.stack.imgur.com/17Xx7.png)
## Proposal
A widget that **allows its children to be positioned based on the siblings** at the same time that **that widget gets its height based not only on its children but also on its position** since it matters. Something like this (there may be a cleaner way and names may be awfull but this is what I came up with, I hope you get the idea)

```Dart
Constraint(
  children: [
    Constrained(
      id: 1,
      child: A(),
      // Default constraints topLeft to parent
    ),
    Constrained(
      id: 2,
      child: B(),
      // Allows multiple constraints, if No ID is constrained to its parent
      constraints: [
        Constraint(Constraint.centerToBottom, id: 1),
      ],
    )
  ],
)
```

### Notes
Correct me if I'm wrong but I think this would still be *single pass* since you ask for their height once, and their position once, that's enough to get the parent's height.

I may be wrong and there may be a widget that allows this but either I didn't find any or I don't know how to achieve this with the ones I've tried without ugly workarounds with "fake" widgets to center **B** (also not allways would be possible since **B** could be a huge Widget with a lot of logic and to have many of it invisible wouldn't be an option for those cases) and also, with the "fake" widgets the parent height will be bigger than the children's height less the shared area.

Widgets that I've tried to solve this with: `Column`, `Row`, `Expanded`, `Align`, `Stack` and `Positioned`, `FractionallySizedBox`, `CustomMultiChildLayout`, `LayoutBuilder`.

Thanks for your time.

**Comment by danagbemava-nc** [comment]
Hi @Gilga89, can you provide some code samples of what you tried and how the existing widgets didn't help you achieve your use case?

**Comment by Gilga89** [comment]
> Hi @Gilga89, can you provide some code samples of what you tried and how the existing widgets didn't help you achieve your use case?

No, I think there is no current way of achieving the image above without a fixed size for **B** widget, which is why I asked for this, I don't have the examples any more, but how could you position **B** like that without knowing its height beforehand, and beeing a height that could change with its content.

**Comment by danagbemava-nc** [comment]
Can you take a look at the sample below to see if it suits your use case? 

<details>
<summary>code sample</summary>

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key, required this.title}) : super(key: key);

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int _counter = 0;

  void _incrementCounter() {
    setState(() {
      _counter++;
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            Container(
              height: 150,
              width: 400,
              color: Colors.blue,
              child: Center(
                child: Text('Widget A'),
              ),
            ),
            Transform.translate(
              offset: Offset(0, -20),
              child: Container(
                width: 150,
                padding: const EdgeInsets.all(16),
                color: Colors.green,
                child: Center(
                  child: Text('Widget B'),
                ),
              ),
            ),
          ],
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: _incrementCounter,
        tooltip: 'Increment',
        child: const Icon(Icons.add),
      ),
    );
  }
}

```
</details>

**Comment by gilgamesharcadia** [comment]
> Can you take a look at the sample below to see if it suits your use case?
> 
> code sample
> ```dart
> import 'package:flutter/material.dart';
> 
> void main() {
>   runApp(const MyApp());
> }
> 
> class MyApp extends StatelessWidget {
>   const MyApp({Key? key}) : super(key: key);
> 
>   @override
>   Widget build(BuildContext context) {
>     return MaterialApp(
>       title: 'Flutter Demo',
>       theme: ThemeData(
>         primarySwatch: Colors.blue,
>       ),
>       home: const MyHomePage(title: 'Flutter Demo Home Page'),
>     );
>   }
> }
> 
> class MyHomePage extends StatefulWidget {
>   const MyHomePage({Key? key, required this.title}) : super(key: key);
> 
>   final String title;
> 
>   @override
>   State<MyHomePage> createState() => _MyHomePageState();
> }
> 
> class _MyHomePageState extends State<MyHomePage> {
>   int _counter = 0;
> 
>   void _incrementCounter() {
>     setState(() {
>       _counter++;
>     });
>   }
> 
>   @override
>   Widget build(BuildContext context) {
>     return Scaffold(
>       appBar: AppBar(
>         title: Text(widget.title),
>       ),
>       body: Center(
>         child: Column(
>           mainAxisAlignment: MainAxisAlignment.center,
>           children: <Widget>[
>             Container(
>               height: 150,
>               width: 400,
>               color: Colors.blue,
>               child: Center(
>                 child: Text('Widget A'),
>               ),
>             ),
>             Transform.translate(
>               offset: Offset(0, -20),
>               child: Container(
>                 width: 150,
>                 padding: const EdgeInsets.all(16),
>                 color: Colors.green,
>                 child: Center(
>                   child: Text('Widget B'),
>                 ),
>               ),
>             ),
>           ],
>         ),
>       ),
>       floatingActionButton: FloatingActionButton(
>         onPressed: _incrementCounter,
>         tooltip: 'Increment',
>         child: const Icon(Icons.add),
>       ),
>     );
>   }
> }
> ```

No, it doesn't, you're using a fixed offset number (-20). I need this to work for dynamic content height, imagine you want to allow B to be as big as its text, if you replace `Text('Widget B')` for `Text('Widget B\nNew line')` you'll see how is not positioned as I wanted it to be.

**Comment by Gilga89** [comment]
> Can you take a look at the sample below to see if it suits your use case?

Hi @danagbemava-nc, no, it doesn't fit my case, you're using a fixed offset number (-20). I need this to work for **dynamic content height**, imagine you want to allow **B** to be as big as its text, if you replace `Text('Widget B')` for `Text('Widget B\nNew line')` you'll see how is not positioned as I wanted it to be.

**Comment by danagbemava-nc** [comment]
Labeling for further input from the team

**Comment by Gilga89** [comment]
> I think you can implement this with a CustomMultiChildLayout. 

From the `CustomMultiChildLayout` docs: "A widget that uses a delegate to size and position multiple children. The delegate can determine the layout constraints for each child and can decide where to position each child. The delegate can also determine the size of the parent, but **the size of the parent cannot depend on the sizes of the children.**"

> If that doesn't work, you can always write a custom RenderObject, that does exactly what you want.

I guess, but that seems like a lot of work for such a simple and common task, I don't even know where to begin with...
> I currently don't see enough of a use case for this to have this as a specialized widget in the framework.

Actually this example I made is the latest case but not the only one in which I needed this, and as you can see is a pretty basic one. A "UI design" search on Google could show you more examples in which a constraints system is needed. For what is worth I have a team of native Android, iOS and React Native developers, I made them switch to Flutter and we came across this issue several times, and every time they asked me I had to either change the UI design or use ugly and not all around fixes.

Every time I look for something like this I found a fixed number as an answer, and that makes the UI not responsive... This is a fairly simple task to do and its built in in the Android and iOS native frameworks, even the new Jetpack Compose have constraints too (it's also a declarative framework and have the equivalent of `Row`, `Column` and `Stack` and still they did see a enough of a use case for this so they created a `ConstraintLayout`)
